File: labs/packages/python-sdk/schlep_engine/api/ml_pipeline.py
"""
Machine Learning Pipeline API for Schlep-engine SDK
"""

from typing import Any, Dict, List, Optional, Union
from pathlib import Path

from .base import BaseAPI
from ..models.ml import (
    MLPipelineConfig, MLPipelineResult, TrainingJob, PredictionRequest,
    PredictionResult, ModelInfo, ModelMetrics, MLTaskType, ModelType
)
from ..models.common import APIResponse


class MLPipelineAPI(BaseAPI):
    """
    Machine Learning Pipeline API client.
    
    Provides methods for creating, training, and managing ML models and pipelines.
    """
    
    def __init__(self, client):
        """Initialize ML pipeline API."""
        super().__init__(client)
        self.base_path = "/ml"
    
    async def create_pipeline(self, config: MLPipelineConfig) -> Dict[str, Any]:
        """
        Create a new ML pipeline.
        
        Args:
            config: Pipeline configuration
            
        Returns:
            Created pipeline information
        """
        response = await self._post("/pipelines", json=config.to_dict())
        return response.get("data", response)
    
    async def get_pipeline(self, pipeline_id: str) -> MLPipelineConfig:
        """
        Get pipeline configuration.
        
        Args:
            pipeline_id: Pipeline ID
            
        Returns:
            Pipeline configuration
        """
        response = await self._get(f"/pipelines/{pipeline_id}")
        pipeline_data = response.get("data", response)
        
        return MLPipelineConfig(
            name=pipeline_data["name"],
            task_type=MLTaskType(pipeline_data["task_type"]),
            model_type=ModelType(pipeline_data["model_type"]),
            target_column=pipeline_data["target_column"],
            **{k: v for k, v in pipeline_data.items() 
               if k not in ["name", "task_type", "model_type", "target_column"]}
        )
    
    async def update_pipeline(
        self,
        pipeline_id: str,
        config: MLPipelineConfig
    ) -> Dict[str, Any]:
        """
        Update pipeline configuration.
        
        Args:
            pipeline_id: Pipeline ID
            config: Updated configuration
            
        Returns:
            Updated pipeline information
        """
        return await self._put(f"/pipelines/{pipeline_id}", json=config.to_dict())
    
    async def delete_pipeline(self, pipeline_id: str) -> Dict[str, Any]:
        """
        Delete a pipeline.
        
        Args:
            pipeline_id: Pipeline ID
            
        Returns:
            Deletion confirmation
        """
        return await self._delete(f"/pipelines/{pipeline_id}")
    
    async def list_pipelines(
        self,
        page: int = 1,
        page_size: int = 20,
        task_type: Optional[MLTaskType] = None,
        model_type: Optional[ModelType] = None
    ) -> APIResponse[List[Dict[str, Any]]]:
        """
        List ML pipelines.
        
        Args:
            page: Page number
            page_size: Items per page
            task_type: Filter by task type
            model_type: Filter by model type
            
        Returns:
            Paginated list of pipelines
        """
        params = self._build_query_params(
            page=page,
            page_size=page_size,
            task_type=task_type.value if task_type else None,
            model_type=model_type.value if model_type else None
        )
        
        return await self._paginated_request("/pipelines", params)
    
    async def train_pipeline(
        self,
        pipeline_id: str,
        training_data_path: Optional[str] = None,
        parameters: Optional[Dict[str, Any]] = None
    ) -> TrainingJob:
        """
        Start training a pipeline.
        
        Args:
            pipeline_id: Pipeline ID
            training_data_path: Path to training data (overrides pipeline config)
            parameters: Additional training parameters
            
        Returns:
            Training job information
        """
        data = {
            "pipeline_id": pipeline_id,
            "training_data_path": training_data_path,
            "parameters": parameters or {}
        }
        
        response = await self._post("/train", json=data)
        training_data = response.get("data", response)
        
        return TrainingJob(
            job_id=training_data["job_id"],
            pipeline_id=training_data["pipeline_id"],
            status=training_data["status"],
            **{k: v for k, v in training_data.items() 
               if k not in ["job_id", "pipeline_id", "status"]}
        )
    
    async def get_training_job(self, job_id: str) -> TrainingJob:
        """
        Get training job status and details.
        
        Args:
            job_id: Training job ID
            
        Returns:
            Training job information
        """
        response = await self._get(f"/training/{job_id}")
        job_data = response.get("data", response)
        
        # Parse metrics if present
        metrics = None
        if job_data.get("metrics"):
            metrics = ModelMetrics(**job_data["metrics"])
        
        return TrainingJob(
            job_id=job_data["job_id"],
            pipeline_id=job_data["pipeline_id"],
            status=job_data["status"],
            metrics=metrics,
            **{k: v for k, v in job_data.items() 
               if k not in ["job_id", "pipeline_id", "status", "metrics"]}
        )
    
    async def cancel_training(self, job_id: str) -> Dict[str, Any]:
        """
        Cancel a training job.
        
        Args:
            job_id: Training job ID
            
        Returns:
            Cancellation confirmation
        """
        return await self._post(f"/training/{job_id}/cancel")
    
    async def get_training_logs(
        self,
        job_id: str,
        lines: Optional[int] = None
    ) -> List[str]:
        """
        Get training logs.
        
        Args:
            job_id: Training job ID
            lines: Number of recent lines to fetch
            
        Returns:
            List of log lines
        """
        params = self._build_query_params(lines=lines)
        response = await self._get(f"/training/{job_id}/logs", params=params)
        return response.get("data", response).get("logs", [])
    
    async def list_training_jobs(
        self,
        page: int = 1,
        page_size: int = 20,
        status: Optional[str] = None,
        pipeline_id: Optional[str] = None
    ) -> APIResponse[List[TrainingJob]]:
        """
        List training jobs.
        
        Args:
            page: Page number
            page_size: Items per page
            status: Filter by job status
            pipeline_id: Filter by pipeline ID
            
        Returns:
            Paginated list of training jobs
        """
        params = self._build_query_params(
            page=page,
            page_size=page_size,
            status=status,
            pipeline_id=pipeline_id
        )
        
        return await self._paginated_request("/training", params, TrainingJob)
    
    async def predict(
        self,
        model_id: str,
        input_data: Union[Dict[str, Any], List[Dict[str, Any]]],
        return_probabilities: bool = False,
        explain_predictions: bool = False
    ) -> PredictionResult:
        """
        Make predictions with a trained model.
        
        Args:
            model_id: Model ID
            input_data: Input data for prediction
            return_probabilities: Whether to return prediction probabilities
            explain_predictions: Whether to include prediction explanations
            
        Returns:
            Prediction results
        """
        request = PredictionRequest(
            model_id=model_id,
            input_data=input_data,
            return_probabilities=return_probabilities,
            explain_predictions=explain_predictions
        )
        
        response = await self._post("/predict", json=request.to_dict())
        result_data = response.get("data", response)
        
        return PredictionResult(**result_data)
    
    async def batch_predict(
        self,
        model_id: str,
        data_path: str,
        output_path: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Run batch predictions on a dataset.
        
        Args:
            model_id: Model ID
            data_path: Path to input data file
            output_path: Path for output predictions
            
        Returns:
            Batch prediction job information
        """
        data = {
            "model_id": model_id,
            "data_path": data_path,
            "output_path": output_path
        }
        
        return await self._post("/predict/batch", json=data)
    
    async def get_model(self, model_id: str) -> ModelInfo:
        """
        Get model information.
        
        Args:
            model_id: Model ID
            
        Returns:
            Model information
        """
        response = await self._get(f"/models/{model_id}")
        model_data = response.get("data", response)
        
        # Parse metrics if present
        metrics = None
        if model_data.get("metrics"):
            metrics = ModelMetrics(**model_data["metrics"])
        
        return ModelInfo(
            model_id=model_data["model_id"],
            name=model_data["name"],
            model_type=ModelType(model_data["model_type"]),
            task_type=MLTaskType(model_data["task_type"]),
            version=model_data["version"],
            status=model_data["status"],
            metrics=metrics,
            **{k: v for k, v in model_data.items() 
               if k not in ["model_id", "name", "model_type", "task_type", "version", "status", "metrics"]}
        )
    
    async def list_models(
        self,
        page: int = 1,
        page_size: int = 20,
        task_type: Optional[MLTaskType] = None,
        model_type: Optional[ModelType] = None
    ) -> APIResponse[List[ModelInfo]]:
        """
        List trained models.
        
        Args:
            page: Page number
            page_size: Items per page
            task_type: Filter by task type
            model_type: Filter by model type
            
        Returns:
            Paginated list of models
        """
        params = self._build_query_params(
            page=page,
            page_size=page_size,
            task_type=task_type.value if task_type else None,
            model_type=model_type.value if model_type else None
        )
        
        return await self._paginated_request("/models", params, ModelInfo)
    
    async def delete_model(self, model_id: str) -> Dict[str, Any]:
        """
        Delete a model.
        
        Args:
            model_id: Model ID
            
        Returns:
            Deletion confirmation
        """
        return await self._delete(f"/models/{model_id}")
    
    async def download_model(self, model_id: str, output_path: str) -> str:
        """
        Download a trained model.
        
        Args:
            model_id: Model ID
            output_path: Local path to save the model
            
        Returns:
            Downloaded file path
        """
        # This would need to be implemented based on the actual API
        # For now, return the API endpoint for model download
        download_url = f"{self.client.base_url}/api/v1/ml/models/{model_id}/download"
        
        # In a real implementation, you'd download the file here
        # For now, we'll just return the URL
        return download_url
    
    async def get_model_metrics(self, model_id: str) -> ModelMetrics:
        """
        Get detailed model metrics.
        
        Args:
            model_id: Model ID
            
        Returns:
            Model metrics
        """
        response = await self._get(f"/models/{model_id}/metrics")
        metrics_data = response.get("data", response)
        
        return ModelMetrics(**metrics_data)

File: labs/packages/javascript-sdk/src/api/ml-pipeline.ts
/**
 * ML Pipeline API for Schlep-engine JavaScript SDK
 */

import { BaseAPI } from './base';
import {
  MLPipelineConfig,
  TrainingJob,
  MLPipelineResult,
  PredictionRequest,
  PredictionResult,
  ModelInfo,
  ModelMetrics,
  MLTaskType,
  ModelType,
  TrainingStatus,
  HyperparameterConfig,
  FeatureEngineering,
  CrossValidationConfig,
  DeploymentConfig,
  BatchPredictionJob,
  ModelMonitoring,
  AutoMLConfig,
  ModelRegistryEntry,
  ExplainabilityRequest,
  ExplainabilityResult
} from '../types/ml';
import { APIResponse } from '../types/common';
import { ProgressCallback } from '../types/common';

/**
 * ML Pipeline API client
 * Provides methods for machine learning model training, deployment, and prediction
 */
export class MLPipelineAPI extends BaseAPI {
  constructor(client: any) {
    super(client);
    this.basePath = '/ml';
  }

  /**
   * Create ML pipeline
   */
  async createPipeline(config: MLPipelineConfig): Promise<APIResponse<{ pipeline_id: string }>> {
    this.validateRequired(config, ['name', 'task_type', 'model_type', 'target_column']);
    
    return this.post<{ pipeline_id: string }>('/pipelines', {
      params: config
    });
  }

  /**
   * Get ML pipeline configuration
   */
  async getPipeline(pipelineId: string): Promise<APIResponse<MLPipelineConfig & { pipeline_id: string }>> {
    this.validateRequired({ pipelineId }, ['pipelineId']);
    return this.get<MLPipelineConfig & { pipeline_id: string }>(`/pipelines/${pipelineId}`);
  }

  /**
   * Update ML pipeline configuration
   */
  async updatePipeline(
    pipelineId: string, 
    updates: Partial<MLPipelineConfig>
  ): Promise<APIResponse<MLPipelineConfig & { pipeline_id: string }>> {
    this.validateRequired({ pipelineId }, ['pipelineId']);
    return this.patch<MLPipelineConfig & { pipeline_id: string }>(`/pipelines/${pipelineId}`, {
      params: updates
    });
  }

  /**
   * Delete ML pipeline
   */
  async deletePipeline(pipelineId: string): Promise<APIResponse<{ message: string }>> {
    this.validateRequired({ pipelineId }, ['pipelineId']);
    return this.delete<{ message: string }>(`/pipelines/${pipelineId}`);
  }

  /**
   * List ML pipelines
   */
  async listPipelines(params?: {
    page?: number;
    pageSize?: number;
    taskType?: MLTaskType;
    modelType?: ModelType;
  }): Promise<APIResponse<Array<MLPipelineConfig & { pipeline_id: string }>>> {
    return this.paginatedRequest('/pipelines', params);
  }

  /**
   * Train ML model
   */
  async trainModel(
    pipelineId: string,
    trainingData: File | string,
    options: {
      validationData?: File | string;
      testData?: File | string;
      hyperparameterConfig?: HyperparameterConfig;
      featureEngineering?: FeatureEngineering;
      crossValidation?: CrossValidationConfig;
      onProgress?: ProgressCallback;
    } = {}
  ): Promise<APIResponse<TrainingJob>> {
    this.validateRequired({ pipelineId }, ['pipelineId']);

    let params: Record<string, unknown> = {
      pipeline_id: pipelineId,
      hyperparameter_config: options.hyperparameterConfig,
      feature_engineering: options.featureEngineering,
      cross_validation: options.crossValidation
    };

    // Handle training data
    if (typeof trainingData === 'string') {
      params.training_data_url = trainingData;
    } else {
      // Upload training file
      const uploadResponse = await this.uploadFile('/upload', trainingData, {
        data_type: 'training'
      }, options.onProgress);
      
      if (!this.isSuccess(uploadResponse)) {
        throw new Error(`Training data upload failed: ${this.getErrorMessage(uploadResponse)}`);
      }
      
      params.training_data_path = uploadResponse.data.url;
    }

    // Handle validation data if provided
    if (options.validationData) {
      if (typeof options.validationData === 'string') {
        params.validation_data_url = options.validationData;
      } else {
        const validationUpload = await this.uploadFile('/upload', options.validationData, {
          data_type: 'validation'
        });
        if (this.isSuccess(validationUpload)) {
          params.validation_data_path = validationUpload.data.url;
        }
      }
    }

    // Handle test data if provided
    if (options.testData) {
      if (typeof options.testData === 'string') {
        params.test_data_url = options.testData;
      } else {
        const testUpload = await this.uploadFile('/upload', options.testData, {
          data_type: 'test'
        });
        if (this.isSuccess(testUpload)) {
          params.test_data_path = testUpload.data.url;
        }
      }
    }

    return this.post<TrainingJob>('/train', { params });
  }

  /**
   * Get training job status
   */
  async getTrainingStatus(jobId: string): Promise<APIResponse<TrainingJob>> {
    this.validateRequired({ jobId }, ['jobId']);
    return this.get<TrainingJob>(`/train/${jobId}`);
  }

  /**
   * Wait for training completion
   */
  async waitForTraining(
    jobId: string,
    pollInterval = 5000,
    maxWaitTime = 3600000 // 1 hour
  ): Promise<APIResponse<TrainingJob>> {
    return super.waitForJob<TrainingJob>(
      jobId,
      `/ml/train/${jobId}`,
      pollInterval,
      maxWaitTime
    );
  }

  /**
   * Cancel training job
   */
  async cancelTraining(jobId: string): Promise<APIResponse<{ message: string }>> {
    this.validateRequired({ jobId }, ['jobId']);
    return this.delete<{ message: string }>(`/train/${jobId}`);
  }

  /**
   * Get training job results
   */
  async getTrainingResults(jobId: string): Promise<APIResponse<MLPipelineResult>> {
    this.validateRequired({ jobId }, ['jobId']);
    return this.get<MLPipelineResult>(`/train/${jobId}/results`);
  }

  /**
   * Make prediction
   */
  async predict(request: PredictionRequest): Promise<APIResponse<PredictionResult>> {
    this.validateRequired(request, ['model_id', 'input_data']);
    return this.post<PredictionResult>('/predict', { params: request });
  }

  /**
   * Make batch predictions
   */
  async batchPredict(
    modelId: string,
    inputData: File | string,
    options: {
      outputPath?: string;
      returnProbabilities?: boolean;
      explainPredictions?: boolean;
      onProgress?: ProgressCallback;
    } = {}
  ): Promise<APIResponse<BatchPredictionJob>> {
    this.validateRequired({ modelId }, ['modelId']);

    let params: Record<string, unknown> = {
      model_id: modelId,
      output_path: options.outputPath,
      return_probabilities: options.returnProbabilities,
      explain_predictions: options.explainPredictions
    };

    if (typeof inputData === 'string') {
      params.input_url = inputData;
    } else {
      const uploadResponse = await this.uploadFile('/upload', inputData, {
        data_type: 'prediction_input'
      }, options.onProgress);
      
      if (!this.isSuccess(uploadResponse)) {
        throw new Error(`Input data upload failed: ${this.getErrorMessage(uploadResponse)}`);
      }
      
      params.input_path = uploadResponse.data.url;
    }

    return this.post<BatchPredictionJob>('/batch-predict', { params });
  }

  /**
   * Get batch prediction job status
   */
  async getBatchPredictionStatus(jobId: string): Promise<APIResponse<BatchPredictionJob>> {
    this.validateRequired({ jobId }, ['jobId']);
    return this.get<BatchPredictionJob>(`/batch-predict/${jobId}`);
  }

  /**
   * List models
   */
  async listModels(params?: {
    page?: number;
    pageSize?: number;
    taskType?: MLTaskType;
    modelType?: ModelType;
    status?: string;
  }): Promise<APIResponse<ModelInfo[]>> {
    return this.paginatedRequest('/models', params);
  }

  /**
   * Get model information
   */
  async getModel(modelId: string): Promise<APIResponse<ModelInfo>> {
    this.validateRequired({ modelId }, ['modelId']);
    return this.get<ModelInfo>(`/models/${modelId}`);
  }

  /**
   * Update model metadata
   */
  async updateModel(
    modelId: string,
    updates: {
      name?: string;
      description?: string;
      tags?: string[];
    }
  ): Promise<APIResponse<ModelInfo>> {
    this.validateRequired({ modelId }, ['modelId']);
    return this.patch<ModelInfo>(`/models/${modelId}`, { params: updates });
  }

  /**
   * Delete model
   */
  async deleteModel(modelId: string): Promise<APIResponse<{ message: string }>> {
    this.validateRequired({ modelId }, ['modelId']);
    return this.delete<{ message: string }>(`/models/${modelId}`);
  }

  /**
   * Deploy model
   */
  async deployModel(config: DeploymentConfig): Promise<APIResponse<{
    deployment_id: string;
    endpoint_url: string;
    status: string;
  }>> {
    this.validateRequired(config, ['deployment_name', 'model_id']);
    return this.post<{
      deployment_id: string;
      endpoint_url: string;
      status: string;
    }>('/deploy', { params: config });
  }

  /**
   * Get deployment status
   */
  async getDeploymentStatus(deploymentId: string): Promise<APIResponse<{
    deployment_id: string;
    model_id: string;
    status: string;
    endpoint_url: string;
    health_status: string;
    metrics?: ModelMonitoring;
  }>> {
    this.validateRequired({ deploymentId }, ['deploymentId']);
    return this.get(`/deployments/${deploymentId}`);
  }

  /**
   * Update deployment
   */
  async updateDeployment(
    deploymentId: string,
    updates: {
      minInstances?: number;
      maxInstances?: number;
      instanceType?: string;
    }
  ): Promise<APIResponse<{ message: string }>> {
    this.validateRequired({ deploymentId }, ['deploymentId']);
    return this.patch<{ message: string }>(`/deployments/${deploymentId}`, { params: updates });
  }

  /**
   * Undeploy model
   */
  async undeployModel(deploymentId: string): Promise<APIResponse<{ message: string }>> {
    this.validateRequired({ deploymentId }, ['deploymentId']);
    return this.delete<{ message: string }>(`/deployments/${deploymentId}`);
  }

  /**
   * List deployments
   */
  async listDeployments(params?: {
    page?: number;
    pageSize?: number;
    status?: string;
    environment?: string;
  }): Promise<APIResponse<Array<{
    deployment_id: string;
    deployment_name: string;
    model_id: string;
    status: string;
    environment: string;
    created_at: string;
  }>>> {
    return this.paginatedRequest('/deployments', params);
  }

  /**
   * Get model metrics
   */
  async getModelMetrics(modelId: string): Promise<APIResponse<ModelMetrics>> {
    this.validateRequired({ modelId }, ['modelId']);
    return this.get<ModelMetrics>(`/models/${modelId}/metrics`);
  }

  /**
   * Explain model predictions
   */
  async explainPrediction(request: ExplainabilityRequest): Promise<APIResponse<ExplainabilityResult>> {
    this.validateRequired(request, ['model_id', 'input_data', 'method']);
    return this.post<ExplainabilityResult>('/explain', { params: request });
  }

  /**
   * Run AutoML
   */
  async runAutoML(
    config: AutoMLConfig,
    trainingData: File | string,
    options: {
      validationData?: File | string;
      onProgress?: ProgressCallback;
    } = {}
  ): Promise<APIResponse<TrainingJob>> {
    this.validateRequired(config, ['task_type', 'target_column']);

    let params: Record<string, unknown> = { ...config };

    // Handle training data
    if (typeof trainingData === 'string') {
      params.training_data_path = trainingData;
    } else {
      const uploadResponse = await this.uploadFile('/upload', trainingData, {
        data_type: 'training'
      }, options.onProgress);
      
      if (!this.isSuccess(uploadResponse)) {
        throw new Error(`Training data upload failed: ${this.getErrorMessage(uploadResponse)}`);
      }
      
      params.training_data_path = uploadResponse.data.url;
    }

    // Handle validation data if provided
    if (options.validationData) {
      if (typeof options.validationData === 'string') {
        params.validation_data_path = options.validationData;
      } else {
        const validationUpload = await this.uploadFile('/upload', options.validationData, {
          data_type: 'validation'
        });
        if (this.isSuccess(validationUpload)) {
          params.validation_data_path = validationUpload.data.url;
        }
      }
    }

    return this.post<TrainingJob>('/automl', { params });
  }

  /**
   * Get feature importance
   */
  async getFeatureImportance(modelId: string): Promise<APIResponse<Record<string, number>>> {
    this.validateRequired({ modelId }, ['modelId']);
    return this.get<Record<string, number>>(`/models/${modelId}/feature-importance`);
  }

  /**
   * Download model artifact
   */
  getModelDownloadUrl(modelId: string, artifactType = 'model'): string {
    this.validateRequired({ modelId }, ['modelId']);
    return this.buildDownloadUrl(`/models/${modelId}/download`, { type: artifactType });
  }

  /**
   * Register model in model registry
   */
  async registerModel(entry: Omit<ModelRegistryEntry, 'created_at' | 'updated_at'>): Promise<APIResponse<ModelRegistryEntry>> {
    this.validateRequired(entry, ['model_id', 'name', 'version']);
    return this.post<ModelRegistryEntry>('/registry/models', { params: entry });
  }

  /**
   * Get registered model
   */
  async getRegisteredModel(modelId: string, version?: string): Promise<APIResponse<ModelRegistryEntry>> {
    this.validateRequired({ modelId }, ['modelId']);
    const path = version ? `/registry/models/${modelId}/versions/${version}` : `/registry/models/${modelId}`;
    return this.get<ModelRegistryEntry>(path);
  }

  /**
   * Update model stage in registry
   */
  async updateModelStage(
    modelId: string,
    version: string,
    stage: 'development' | 'staging' | 'production' | 'archived'
  ): Promise<APIResponse<ModelRegistryEntry>> {
    this.validateRequired({ modelId, version, stage }, ['modelId', 'version', 'stage']);
    return this.patch<ModelRegistryEntry>(`/registry/models/${modelId}/versions/${version}`, {
      params: { stage }
    });
  }

  /**
   * List training jobs
   */
  async listTrainingJobs(params?: {
    page?: number;
    pageSize?: number;
    status?: TrainingStatus;
    pipelineId?: string;
  }): Promise<APIResponse<TrainingJob[]>> {
    return this.paginatedRequest('/train', params);
  }

  /**
   * Get ML framework capabilities
   */
  async getCapabilities(): Promise<APIResponse<{
    supported_tasks: MLTaskType[];
    supported_models: ModelType[];
    supported_frameworks: string[];
    max_features: number;
    max_training_time: number;
  }>> {
    return this.get('/capabilities');
  }
}

export default MLPipelineAPI;